NEAT Logic BB, first merge

kristahi opened the pull request:
At UiO we have been working on implementing the NEAT Logic BB, and here are the first results of this work.

Our work has been guided by D1.2 sections 3.2 and 3.3, as well as [draft-ietf-taps-transports-usage-00](https://tools.ietf.org/html/draft-ietf-taps-transports-usage-00) (we're aware of its UDP sibling, waiting for UDP code before we integrate that).

## What's in
The API now conforms to D1.2 sect. 3.2 in it's entirety, although we have not had time to implement all the features yet.

All callbacks work, although the NETWORK_STATUS_CHANGE event is only functional for SCTP at this time. 
This should close #27 , with the addition of `on_close` (AKA CLOSE from D1.2)

We also added the primitives and events from sect. 3.3.1 - 3.3.3 to the API, but they are not functional at this time.

## What's out
Unimplemented features cause return values of NEAT_ERROR_UNABLE. Currently incomplete:

- OPEN with specification of local name to bind to.
- OPEN, WRITE and READ multi-streaming support.
- CHANGE_TIMEOUT for TCP works iff. there is TCP User Timeout support in the kernel (only Linux)

Property-related changes are not part of this PR. We are working on transitioning that to a JSON-based format suitable for interfacing with the Policy Manager, and in line with the discussion in issue #81 (definition of actual properties of course needs further agreement, we are simply laying some groundwork for the implementation).

## SCTP/TCP inconsistensies
Furthermore, several important features work differently across TCP and SCTP. For example, the ABORT callback is not called when running over TCP; this is per the specification, but strikes us as problematic from the application programmer's point of view. We are working on proposals to even this out, so application programmers can rely on this callback (we already did this for the CLOSE event).

## Documentation etc.

The documentation in the docs subdirectory has been updated, and we have tried to stick to the existing style of function calls and so on. The client example has received some new callback handlers.

Sorry for the quite large PR, we will try to keep them smaller and more focused as we iterate. We're eager to hear your feedback and criticisms (and fixing it!).

tuexen commented on neat_internal.h:
```diff
@@ -30,6 +30,14 @@
 #define NEAT_MAX_NUM_PROTO 4
 
+#if defined(USRSCTP_SUPPORT) || defined(HAVE_NETINET_SCTP_H)
+#define SCTP_SUPPORTED
```
Why not be more specific as the code already is. I don't think we need this constant.

tuexen commented on neat_internal.h:
```diff
@@ -30,6 +30,14 @@
+#define SCTP_SUPPORTED
+#else
+#if defined(__linux__) // Need a further test for Linux SCTP support
+#define SCTP_SUPPORTED
```
This is not always true. Just use more specific #ifdefs as in the rest of the code.

kristahi replied:
Just for the record: going with this suggestion, the SCTP_SUPPORTED constant has been removed. 

Running into some bugs in the kind of oddball configuration of Linux kernel SCTP enabled without having the proper headers, see also #86. My suggestion there is to say that it is a bad configuration and avoid using kernel SCTP at all in that case.

tuexen replied:
OK. I would assume that if someone wants to use SCTP, he has the proper headers installed.
I haven't understood why Linux makes headers an optional install although they provide the functionality always. One problem resulting from this is that the kernel headers used for compiling the kernel might be different from the headers the application sees. Have been running into this in the past. Never understood it.

tuexen commented on neat_core.h:
```diff
@@ -7,6 +7,30 @@
+#define SCTP_CAUSE_RESTART_W_NEWADDR	0x000b
+#define SCTP_CAUSE_USER_INITIATED_ABT	0x000c
+#define SCTP_CAUSE_PROTOCOL_VIOLATION	0x000d
+
```
I think it is a bad idea to define constants in neat which belong to a different namespace. The names of the constants are implementation specific. So I guess we should just no use them. Either define them in the NEAT namespace or just use constants.

kristrev replied:
For constants we seemingly control ourself it makes more sense to group them in an enum than to have lots of defines.

tuexen replied:
We don't control them. The values are actually IANA assigned. However, the names are not specified in any API specification. So the names are actually platform specific and we shouldn't define anything in the `SCTP_` namespace.

kristrev replied:
I would still add en enum and specify explicit variable values than adding too many defines.

tuexen replied:
OK. I don't care about enum versus define. But don't interfere with the name space controlled by including `sctp.h`.

kristahi replied:
This was replaced by `enum neat_sctp_cause_code` in [aa42838025](https://github.com/NEAT-project/neat/pull/85/commits/aa42838025d2d5a13f8386138eb02e36bef227fb), with NEAT_ prefix.

tuexen replied:
Great. Thanks.

tuexen commented on doc/callb.on_aborted.md:
```diff
@@ -0,0 +1,10 @@
+**Availability:** Currently only available when NEAT selects SCTP
+  transport. The specification in D1.2 and
+  draft-ietf-taps-transports-usage-00 does not specify this event for
+  TCP; we are investigating how we might extend this work for TCP.
```
Wouldn't it for TCP make sense to fire it is the TCP connection was reset?

kristahi replied:
Totally agree. The spec says it is undefined for TCP, but I plan to implement something for this in the near future - would like to keep it out of this PR.

tuexen replied:
OK

tuexen commented on examples/client_timeout_test.c:
```diff
@@ -0,0 +1,438 @@
+#include <stdlib.h>
```
Would it make sense to just extend `client.c`?

oystedal replied:
In the long term, the answer is probably yes. I did it this way to avoid modifying the existing client, which by itself has no requirements wrt. to timing. I agree that the timeout could be an argument to the existing client instead of a separate program.

oystedal replied:
I have removed this test client for now. I plan on adding support for the timeout call to `client.c` in the future.

tuexen commented on neat_core.c:
```diff
@@ -30,6 +33,9 @@
 #ifdef __linux__
     #include "neat_linux_internal.h"
+#if !defined(HAVE_NETINET_SCTP_H) && !defined(USRSCTP_SUPPORT)
+    #include <linux/sctp.h>
```
Why don't you use netinet/sctp.h?

kristahi replied:
Was missing that header, was not aware of needing the libsctp-dev package to use a kernel feature. Adding a note to README about that, as well.

tuexen replied:
I find this also strange...
I think you can neat without having libsctp-dev installed, SCTP wouldn't just work. So it is only a requirement when you want to use SCTP. But we can just require it and therefore adding it to the required libraries for Linux.

tuexen commented on neat_core.c:
```diff
@@ -465,6 +495,139 @@ static void io_writable(neat_ctx *ctx, neat_flow *flow,
+    neat_error_code outcode;
+
+    // TODO: this translation table is not very good,
+    // should probably improve on it, add more NEAT codes
```
See comment above or below. The constants used are not part of an API.

tuexen commented on neat_core.c:
```diff
@@ -465,6 +495,139 @@ static void io_writable(neat_ctx *ctx, neat_flow *flow,
+    case SCTP_COMM_UP: // Fallthrough:
+    case SCTP_RESTART:
+	// TODO: might want to "translate" the state codes to a NEAT code.
+	neat_notify_network_changed(flow, sac->sac_state);
```
Is this a good name? I would have expected some notification regarding the change of a path state.

kristahi replied:
You mean `neat_notify_network_changed`, right? This is the dispatch function for the NETWORK_STATUS_CHANGE callback. I could rename it to `neat_notify_network_status_changed`, I suppose. The name is still a bit vague, maybe it should be redefined in D1.2 as well.

tuexen commented on neat_core.c:
```diff
@@ -465,6 +495,139 @@ static void io_writable(neat_ctx *ctx, neat_flow *flow,
+// One is generated per failed message
+// Only FreeBSD and USRSCTP support the new RFC6458 API,
+// hence the ifdefs
+#if defined(__FreeBSD__) || defined (USRSCTP_SUPPORT)
```
So wouldn't it make more sense to test for the existence of the structure and base the #ifdef on that? That allows additional platforms to use the correct code path automatically.

kristahi replied:
Good point! Adding tests to CMake script to check for the structs and using that.

tuexen commented on neat_core.c:
```diff
@@ -465,6 +495,139 @@ static void io_writable(neat_ctx *ctx, neat_flow *flow,
+    case SCTP_ASSOC_CHANGE:
+	handle_sctp_assoc_change(flow, &notfn->sn_assoc_change);
+	break;
+#if defined(__FreeBSD__) || defined (USRSCTP_SUPPORT)
```
I would suggest to base the #ifdef on the existence of the constants, not on the platforms.

tuexen commented on neat_core.c:
```diff
@@ -473,7 +636,7 @@ static void io_writable(neat_ctx *ctx, neat_flow *flow,
                         neat_error_code code)
 {
-#if defined(IPPROTO_SCTP)
+#if defined(SCTP_SUPPORTED)
```
Don't use this constant. Test for something you actually need.

tuexen commented on neat_core.c:
```diff
@@ -486,14 +649,14 @@ static int io_readable(neat_ctx *ctx, neat_flow *flow,
     }
-#if defined(IPPROTO_SCTP)
+#if defined(SCTP_SUPPORTED)
     if ((flow->sockProtocol == IPPROTO_SCTP) &&
```
Test for IPPROTO_SCTP.

kristahi replied:
I disagree here. OSX defines IPPROTO_SCTP, but does not have support for it - this code path is useless then. Replacing with
`defined(HAVE_NETINET_SCTP_H) || defined(USRSCTP_SUPPORT)` for now.

tuexen replied:
But it would compile. And when you load an NKE to add SCTP support it would work...
The general rule of the #ifdef mess I used was to compile in what is possible to compile in...

kristahi replied:
Turns out it won't compile, since MSG_NOTIFICATION isn't defined. If I ifdef that out, the implementation will be broken if a module is loaded (no events). There's probably other breakage from other code paths being compiled out, too

Not sure how to best enable that use case, really. Any advice?

tuexen replied:
Correct. When installing the NKE, if also providing a version of sys/socket.h, which contains the definition of
MSG_NOTIFICATION. I was able to convince Apple to include IPPROTO_SCTP, but failed in MSG_NOTIFICATION...
Suggestion:
#if defined(IPPROTO_SCTP) && defined(MSG_NOTIFICATION)

kristahi replied:
Ok, that does make sense for this particular code path. In the bigger picture, I'm a bit worried that things will still be broken elsewhere because of assumptions around what to do when SCTP support is detected or not.

Is there any consensus on how this should work? I see for example that on OSX, HE thinks it can try SCTP, although that isn't really supported. That's different from NetBSD, where it only detects TCP.

Also, what happens if several different SCTP implementations are available?
Right now, all code paths will always prefer USRSCTP if that is compiled in if I understand the code correctly.

tuexen replied:
> On 03 Jun 2016, at 14:47, Kristian A. Hiorth <notifications@github.com> wrote:
> 
> In neat_core.c:
> 
> >      neat_log(NEAT_LOG_DEBUG, "%s", __func__);
> >  
> >      if (!flow->operations || !flow->operations->on_readable) {
> >          return READ_WITH_ERROR;
> >      }
> > -#if defined(IPPROTO_SCTP)
> > +#if defined(SCTP_SUPPORTED)
> >      if ((flow->sockProtocol == IPPROTO_SCTP) &&
> 
> Ok, that does make sense for this particular code path. In the bigger picture, I'm a bit worried that things will still be broken elsewhere because of assumptions around what to do when SCTP support is detected or not.
> 
> Is there any consensus on how this should work? I see for example that on OSX, HE thinks it can try SCTP, although that isn't really supported. That's different from NetBSD, where it only detects TCP.
And that is perfectly fine. HE is for detecting what is working and what not.
> 
> Also, what happens if several different SCTP implementations are available?
Currently you can't compile the userland stack in when a kernel stack is also available.
A fix is on the ToDo list, but has low priority. On FreeBSD I would always prefer
the kernel stack. On Linux it might depend. But we have to deal with this in the
future. Right now our assumption was that the userland stack is only compiled in
when no kernel stack is available.
> Right now, all code paths will always prefer USRSCTP if that is compiled in if I understand the code correctly.
Yes. Right now you can only have one. It is a limitation.

Please note that you can't do native SCTP in userland if the kernel supports SCTP. So the
only option for having both is on Linux and
* ensure that the SCTP module is not loaded (I think it is automatically loaded when opening
  an SCTP socket) and use the userland stack for SCTP/IP and SCTP/UDP/IP.
* use the kernel stack for SCTP/IP and the userland stack for SCTP/UDP/IP.

Best regards
Michael
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

tuexen replied:
And that is perfectly fine. HE is for detecting what is working and what not.

Currently you can't compile the userland stack in, when a kernel stack is also available.
A fix is on the ToDo list, but has low priority. On FreeBSD I would always prefer
the kernel stack. On Linux it might depend. But we have to deal with this in the
future. Right now our assumption was that the userland stack is only compiled in
when no kernel stack is available.

 Right now you can only have one SCTP active in NEAT. This is a limitation.

Please note that you can't do native SCTP in userland if the kernel supports SCTP. So the
only option for having both is on Linux and
* ensure that the SCTP module is not loaded (I think it is automatically loaded when opening
 an SCTP socket) and use the userland stack for SCTP/IP and SCTP/UDP/IP.
* use the kernel stack for SCTP/IP and the userland stack for SCTP/UDP/IP.

tuexen commented on neat_core.c:
```diff
@@ -528,20 +691,15 @@ static int io_readable(neat_ctx *ctx, neat_flow *flow,
 #endif
-	    addr.sin_family = AF_INET;
+	addr.sin_family = AF_INET;
 
```
Why is `addr` needed at all?

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
 }
 
+neat_error_code
+neat_open_localname(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port, const char *localname)
```
Why is this function needed? Isn't `neat_open` good enough?

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+}
+
+neat_error_code
+neat_open_multistream(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port, const char* localname, int count)
```
Do we need a separate function for this? Doesn't is make more sense to have a some form of optional parameter to `neat_open` to indicate the number of streams. If not provided, it is 1.

oystedal replied:
C does not support arguments with default values to my knowledge. Optional arguments could be implemented using a variadic function (... in the parameter list and `va_list` et. al.), but I am opposed to this approach. I'd rather have multiple functions with different names instead of potentially having programs that compile just fine yet displays unusual/undesirable behavior just because optional parameters were specified incorrectly or in the wrong order. I vote for using the little static checking there is in C, at least.

tuexen replied:
I wasn't thinking about variadic functions. I think that we might want to provide a flexible way for providing arguments in multiple places. So I think a consistent way within the neat API should be used.
One way I'm thinking about is to provide a pointer to a sequence of TLVs. It can be NULL, if you caller doesn't want to provide anything, it can provide multiple parameters. Pretty flexible and easy to pass. An alternative would be to use a single pointer and an enumeration type variable to indicate what is provided.
@kristrev Any preference?

oystedal replied:
I have reverted the commits that added these functions for now, until we have decided how to handle optional arguments.

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+#if defined(__FreeBSD__) || defined(__NetBSD__)
+	    // KAH: turns out this was implemented on a feature
+	    // branch, and sadly not merged to mainline
+                TCP_SNDUTO_TIMEOUT,
```
So shouldn't you then fail this function call?

oystedal replied:
Agreed.

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+
+            return NEAT_ERROR_BAD_ARGUMENT;
+    }
+
```
Why is this wrong only on the specified platforms? Wouldn't it make sense to use an unsigned integer as the type for `seconds`?

oystedal replied:
I agree that this should be checked on all platforms.

Using an unsigned integer here wouldn't always catch the case when a user tries to use a negative timeout, with or without intention, if very liberal compiler settings are used. I don't see what sort of application would have a timeout long enough to require that extra bit.

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+            return NEAT_ERROR_BAD_ARGUMENT;
+        }
+
+        unsigned int new_value =
```
Don't declare a variable in the middle of a block.

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+        unsigned int new_value = ((unsigned int)seconds) * 1000 / 4; // seconds -> ms
+        usrsctp_sysctl_set_sctp_heartbeat_interval_default((uint32_t)new_value);
+        usrsctp_sysctl_set_sctp_assoc_rtx_max_default(4);
+#else
```
Why are you changing stack wide settings when the user requested something on a flow. This looks wrong.

oystedal replied:
This was a hacky approach I used to test, before realizing that usersctp wasn't fully supported yet. It's wrong and should be removed.

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+        params.spp_hbinterval = (unsigned int)seconds * 1000 / 4;
+        params.spp_flags |= SPP_HB_ENABLE;
+        params.spp_pathmaxrxt = 4;
+
```
I don't understand these settings. This looks like magic behaviour to the user. I think this is a TCP only feature available only on one platform. So fail this call on other protocols or platforms.

oystedal replied:
The error message is a typo. This block is only executed when SCTP is the protocol in use, and is not related to TCP in any way. It is however quite possibly Linux specific code, so it should be only defined when compiled for Linux.

As for the options themselves, I see that I should have made a comment on them. The approach used to detect a timeout in SCTP is derived from [taps transport usage](https://datatracker.ietf.org/doc/draft-ietf-taps-transports-usage/?include_text=1), page 12. This is achieved by using the heartbeat mechanism with a maximum number of retransmissions. I chose to use 4 heartbeats to avoid firing the timeout just because a single heartbeat was lost. I'm guesstimating 4 as a reasonable number of retransmissions before issuing a timeout, hence dividing the timeout value by 4 to get the heartbeat interval. The number of retransmissions is possibly also application specific, so the code should be changed if we ever modify the number of retransmissions elsewhere.

I agree that this isn't a solid approach. Suggestions are welcome.

tuexen replied:
SCTP does not support killing an association after a specific time. So if the semantic of `neat_change_timeout` is this, fail it for SCTP based associations. Don't provide some magic to the user which doesn't do what the user expects.

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+    }
+#elif defined(HAVE_NETINET_SCTP_H) && defined(__linux__)
+    struct sctp_prim addr;
+    struct sctp_assocparams assocparams;
```
Don't declare variable in the middle of a block.

oystedal replied:
I personally find this to be an unnecessary restriction, and I don't see any technical reason for it unless we're going to support C89/ANSI C as well as C99, which is what the cmake file specifies.

I don't have any major objections, but it would be nice to have a well-defined coding style for the project, which I think is a point that's to be discussed in the plenary in Oslo if I'm not mistaken.

tuexen replied:
UP to now we tried to declare variables only at the beginning of a block. I agree, it would be good if some guidelines are written down. Normally, I just adopt to what is used in a project by looking around, if no formal guidelines are there. I guess you should not see a lot of variables declared in the middle of a block.

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+        ctx->resolver->handle_resolve = set_primary_dest_resolve_cb;
+        neat_getaddrinfo(ctx->resolver, AF_UNSPEC, name, flow->port,
+                protocols, sizeof(*protocols)/sizeof(protocols[0]));
+
```
Isn't that last term a complex way of saying 1.

oystedal replied:
It is, and I prefer to write it that way, should someone modify the `protocols` array for any reason.

tuexen replied:
Not sure if it would make sense to provide there more than one protocol candidate. But does `neat_getaddrinfo` ensure that name is an IP-address? I think the argument `name` must be one...

oystedal replied:
neat_getaddrinfo does not require the `name` argument to be an IP-address.

tuexen replied:
So, how do you enforce it?

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+    }
+#endif
+
+    char dest_addr[INET6_ADDRSTRLEN];
```
Don't declare a variable in the middle of a block.

tuexen commented on neat_core.c:
```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+#endif
+
+    char dest_addr[INET6_ADDRSTRLEN];
+    inet_ntop(results->lh_first->ai_family,
```
Use getnameinfo to support link local addresses

tuexen commented on neat_core.c:
```diff
@@ -1455,6 +1904,72 @@ neat_shutdown_via_kernel(struct neat_ctx *ctx, struct neat_flow *flow)
+#else
+static void neat_sctp_init_events(int sock)
+#endif
+#if defined(USRSCTP_SUPPORT) || defined(__FreeBSD__)
```
Can't you test for `SCTP_EVENT`?

tuexen commented:
I tried to look for SCTP related issues and provided comments. This patch does a lot of things. It should have been splitted in several smaller patches. That would allow for a much simpler review.

oystedal commented:
I agree that a patch like this is too large. In this case we're making additions and modifications that are mostly connected to each other, and all of it is related to conforming with the API as defined in D1.2 section 3.2 and 3.3, which is why we've submitted it as a single patch. We will aim for smaller patches in the future.

oystedal commented:
I believe we have addressed all issues brought forward by your comments @tuexen, the feedback is appreciated. 

Github does not mark all of the diffs above as out of date, but if you look at the files themselves and the commit history on this page, they should all have been fixed now.

If there are no further objections, I suggest that these changes are merged into the master branch.

tuexen commented:
The patch is too big to allow me to say if things are addressed or not. I have to trust you. However, at least the one with setting the primary address seems to be open. It doesn't make sense to provide a name there. It must be a IP-address. I think this should be enforced by the API, which isn't currently the case.

kristahi commented:
> However, at least the one with setting the primary address seems to be open. It doesn't make sense to provide a name there. It must be a IP-address. I think this should be enforced by the API, which isn't currently the case.

Fixed this now.

In the process I moved the function `neat_resolver_check_for_literal` in neat_resolver.c from internal to global linkage, hope nobody minds.

kristahi commented:
Having addressed SCTP service in the absense of netinet/sctp.h (see #86 and [rev 30f7cc6](https://github.com/NEAT-project/neat/pull/85/commits/30f7cc60da6ad2a3fc3cb7913c9393a7dfb9af76))  as well as the problem in `neat_set_primary_addr`, I think we have fixed all outstanding issues.

OK to merge?

tuexen commented:
If you have addressed all issues I raised, you have an OK from my side. But my review focussed mainly on the SCTP aspects...

kristahi commented:
> If you have addressed all issues I raised, you have an OK from my side. But my review focussed mainly on the SCTP aspects...

Thank you so much for your time, by the way! Definitely improved the quality of our code significantly.

The SCTP parts you reviewed really form a large part of the patch, anyway. If nobody raises their hand until noon tomorrow (Wednesday) and commit to reviewing more, I will merge at that time.

tuexen commented:
Makes sense. Assuming that the buildbots are happy (or at least don't complain about issue in your code), go ahead. We can improve the code in the repository...

kristahi commented:
Nobody came forward, so merge went ahead. Buildbots are all green and we don't foresee any problems, but do file issues if you find something.
